amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+    pulse_lengths: list,
+    measure_length: Optional[float] = None,
+    require_2_or_3_between_levels: bool = False,
+):
+    """
+    Convert a list of pulse lengths into a corresponding list of lists
+    with start positions per metrical level.
+    All values (pulse lengths, start positions, and measure_length)
+    are all expressed in terms of quarter length.
+
+    This does not work for ("nonisochronous") pulse streams of varying duration
+    in time signatures like 5/x, 7/x
+    (e.g., the level of 5/4 with dotted/undotted 1/2 notes).
+
+    It is still perfectly fine to use this for the pulse streams
+    within those meters that are regular, equally spaced ("isochronous")
+    (e.g., the 1/4 note level of 5/4).
+
+    The list of pulse lengths is set in decreasing order.
+
+    If `require_2_or_3_between_levels` is True (default), this functions checks that
+    each level is either a 2 or 3 multiple of the next.
+
+    By default, the measure_length is taken by the longest pulse length.
+    Alternatively, this can be user-defined to anything as long as it is
+    1) longer than the longest pulse and
+    2) if `require_2_or_3_between_levels` is True then exactly 2x or 3x longer.
+
+
+    Parameters
+    ----------
+    pulse_lengths
+        Any valid list of pulse lengths, e.g., [4, 2, 1].
+    measure_length
+        Optional. If not provided it's taken to be given by the longest pulse length.
+    require_2_or_3_between_levels
+        Deafults to False. If True, raise a ValueError in the case of this condition not being met.
+
+    Returns
+    -------
+    list
+        Returns a list of lists with start positions by level.
+
+    Examples
+    --------
+
+    >>> qsl = starts_from_pulse_lengths(pulse_lengths=[4, 2, 1, 0.5])
+
+    >>> qsl[0]
+    [0.0, 4.0]
+
+    >>> qsl[1]
+    [0.0, 2.0, 4.0]
+
+    >>> qsl[2]
+    [0.0, 1.0, 2.0, 3.0, 4.0]
+
+    >>> qsl[3]
+    [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0]
+
+    """
+
+    pulse_lengths = sorted(pulse_lengths)[::-1]  # largest number first
+
+    if not measure_length:
+        measure_length = float(pulse_lengths[0])
+
+    else:
+        if pulse_lengths[0] > measure_length:
+            raise ValueError(
+                f"The `pulse_length` {pulse_lengths[0]} is longer than the `measure_length` ({measure_length})."
+            )
+
+    if require_2_or_3_between_levels:
+        for level in range(len(pulse_lengths) - 1):
+            if pulse_lengths[level] / pulse_lengths[level + 1] not in [2, 3]:
+                raise ValueError(
+                    "The proportion between consecutive levels is not 2 or 3 in "
+                    f"this case: {pulse_lengths[level]}:{pulse_lengths[level + 1]}."
+                )
+
+    start_list = []
+
+    for pulse_length in pulse_lengths:
+        starts = starts_from_lengths(measure_length, pulse_length)
+        start_list.append(starts)
+
+    return start_list
+
+
+def coincident_pulse_list(
```

Comment:
I'm a bit nervous that this function relies on floating point equality. Better to implement some kind of tolerance for approximate matching?

Reply:
One to discuss - @rbdannenberg has raised this question for wider repo mattes. Issue.